osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs

```diff
@@ -4,98 +4,281 @@
 #nullable disable
 
 using System;
-using osu.Framework.Utils;
 using osu.Game.Rulesets.Difficulty.Preprocessing;
 using osu.Game.Rulesets.Difficulty.Skills;
 using osu.Game.Rulesets.Mania.Difficulty.Preprocessing;
 using osu.Game.Rulesets.Mods;
 
+/* Rules of Strain Calculation
+ * 1) The current strain value can only depend on past notes
+ * 2) Notes on the same offset may see different strain values, as column order sorting is not guaranteed.
+ * 3) 1st hit object is OMITTED as it acts as a reference for the following note.
+ *
+ * E.g. A:(100ms @ Col 1), B:(200ms @ Col 1), C:(200ms @ Col 2)
+ * The following sequences are possible: B C, C B.
+ * A is omitted because of rule 3.
+ */
+
+/* Previous Note States
+ * When calculating strain by its history, there's these possible states.
+ *
+ * The column describes where the previous note's end time is.
+ * E.g. E3 states that the previous note's end time is on the body (of the current note)
+ *
+ * Impossible states are marked with X, as their head is AFTER our current offset.
+ *
+ * Legend
+ * ┌──────────────┐
+ * │ (State Name) │
+ * │ Prev Note    │
+ * │ This Note    │
+ * └──────────────┘
+ *                 ┌─────────────┬─────────────┬─────────────┬─────────────┬──────────────┐
+ *                 │ Before Head │ On Head     │ On Body     │  On Tail    │ After Tail   │
+ *                 │ (1)         │ (2)         │ (3)         │  (4)        │ (5)          │
+ * ┌───────────────┼─────────────┼─────────────┼─────────────┼─────────────┼──────────────┤
+ * │ Note      (A) │ (A1)        │ (A2)        │             │             │              │
+ * │ Before        │ O           │      O      │      X      │      X      │      X       │
+ * │ Note          │      O      │      O      │             │             │              │
+ * ├───────────────┼─────────────┼─────────────┼─────────────┼─────────────┼──────────────┤
+ * │ Long Note (B) │ (B1)        │ (B2)        │             │             │ (B5)         │
+ * │ Before        │ [==]        │ [====]      │      X      │      X      │ [==========] │
+ * │ Note          │      O      │      O      │             │             │      O       │
```

Comment:
Don't know if just me, but the way this is written is very confusing. There are "on body" and "on tail" columns, but the object being currently considered is a plain note. I think I see what this is angling for here, it's representing a plain note as a zero-duration hold, and as such (2), (3), and (4) are effectively the same thing - but this needs explaining.

Reply:
Yes, unfortunately, the code itself also hints a note as a 0-length hold, which makes sense in retrospect. 
I'll note that beside the table

```cs
// Given a note, startTime == endTime.
double startTime = hitObject.StartTime;
double endTime = hitObject.EndTime;
```